test/e2e/route_metrics_test.go

```diff
@@ -0,0 +1,393 @@
+// For testing the case where IC has only NamespaceLabelSelector then testRS should be set to false and testNS should
+// be set to true. For testing the case where IC has both RouteLabelSelector and NamespaceLabelSelector then both testRS
+// and testNS should be set to true. This function should not be called with both testRS and testNS set to false.
+func testRouteMetricsControllerLabelSelector(t *testing.T, testRS, testNS bool) {
+
+	// Create a new prometheus client for fetching metrics.
+	kubeConfig, err := config.GetConfig()
+	if err != nil {
+		t.Fatalf("failed to get kube config: %s", err)
+	}
+	kubeClient, err := kubernetes.NewForConfig(kubeConfig)
+	if err != nil {
+		t.Fatal(err)
+	}
+	routeClient, err := routev1client.NewForConfig(kubeConfig)
+	if err != nil {
+		t.Fatal(err)
+	}
+	prometheusClient, err := metrics.NewPrometheusClient(context.TODO(), kubeClient, routeClient)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Assign the correct values to the strings depending on the values of testRS and testNS.
+	icNameStr := ""
+	correctLabel := ""
+	incorrectLabel := ""
+	routeNameStr := ""
+	if testRS && testNS {
+		icNameStr = "ic-rs-ns-metrics-test"
+		correctLabel = "rs-ns-foo"
+		incorrectLabel = "rs-ns-bar"
+		routeNameStr = "route-rs-ns-foo-label"
+	} else if testNS {
+		icNameStr = "ic-ns-metrics-test"
+		correctLabel = "ns-foo"
+		incorrectLabel = "ns-bar"
+		routeNameStr = "route-ns-foo-label"
+	} else if testRS {
+		icNameStr = "ic-rs-metrics-test"
+		correctLabel = "rs-foo"
+		incorrectLabel = "rs-bar"
+		routeNameStr = "route-rs-foo-label"
+	} else {
+		t.Fatalf("either of testRS or testNS should be set to true. testRS: %t, testNS: %t", testRS, testNS)
+	}
+
+	// Create an Ingress Controller that can admit our Route.
+	icName := types.NamespacedName{Namespace: operatorNamespace, Name: icNameStr}
+	domain := icName.Name + "." + dnsConfig.Spec.BaseDomain
+	ic := newPrivateController(icName, domain)
+
+	// Add RouteSelector if testRS is set to true.
+	if testRS {
+		ic.Spec.RouteSelector = &metav1.LabelSelector{
+			MatchLabels: map[string]string{
+				"type": correctLabel,
+			},
+		}
+	}
+
+	// Add NamespaceSelector if testNS is set to true.
+	if testNS {
+		ic.Spec.NamespaceSelector = &metav1.LabelSelector{
+			MatchLabels: map[string]string{
+				"type": correctLabel,
+			},
+		}
+	}
+
+	if err := kclient.Create(context.TODO(), ic); err != nil {
+		t.Fatalf("failed to create ingresscontroller: %v", err)
+	}
+
+	// Delete the IC if any error occurs.
+	defer assertIngressControllerDeleted(t, kclient, ic)
+
+	// Store the start time to get the time taken to update metrics after the creation of the IC.
+	startTime := time.Now()
+
+	// Wait for metrics to be added and set to 0.
+	if err := waitForRouteMetricsAddorUpdate(t, prometheusClient, ic.Name, 0); err != nil {
+		t.Fatalf("failed to fetch expected metrics: %v", err)
+	}
```

Comment:
Initializing the metric is a step that will not happen in the real world, so please remove it from your test.

Reply:
Removed this check.

Reply:
@candita Sorry to comment on already reviewed things, but what do you mean "will not happen in the real world"? As far as I can tell, this PR will indeed initialize the metrics for this ingress controller to 0 (or however many routes admitted) upon it's creation. In this case, I thought Arkadeep was confirming the pre-condition that no other routes are admitted and the IC's metrics are not starting in a bogus state. Thoughts?

Reply:
@gcs278 I thought I asked about removing the initialization of the metric to 0, it's not needed.  I guess that was not done after all.

Reply:
Ah I see, I think maybe this was a misunderstanding. Arkadeep was not initializing anything here, he was waiting for the the metrics to be initialized to 0 by the route metric controller, but he wasn't triggering that process.

I'd say this is a desirable check and we should put it back in. @candita let me know if you have any issues with that.

Reply:
@gcs278 added back the check.